### PR TITLE
Add active run ID as trace `mlflow.sourceRun`  tag

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11821?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11821/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11821
```

</p>
</details>

### What changes are proposed in this pull request?

Record the active run ID in `mlflow.sourceRun` tag if the trace is created within the context of a run.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

```
import mlflow
from mlflow.tracking.client import MlflowClient
from mlflow.tracking.fluent import _get_experiment_id

with mlflow.start_run(run_name="run_with_span") as run:
    with mlflow.start_span(name="span") as span:
        span.set_inputs({"x": 1})
        span.set_outputs(10)
    print(f"Run ID: {run.info.run_id}")

> Run ID: d88960eb19c6495c9975b0e149bea699

mlflow.get_traces()[-1].info.tags

> {'mlflow.sourceRun': 'd88960eb19c6495c9975b0e149bea699', ... }
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -875,7 +875,7 @@ def _upload_ended_trace_info(
 
     def _exclude_immutable_tags(self, tags: Dict[str, str]) -> Dict[str, str]:
         """Exclude immutable tags e.g. "mlflow.user" from the given tags."""
-        return {k: v for k, v in tags.items() if not k.startswith("mlflow.")}
```
Comment (B-Step62) [inline mlflow/tracking/client.py:878]: The `mlflow.sourceRun` tag is generated by client side so we can't reject all tags starts with `mlflow.`.

File: tests/tracing/conftest.py [hunk 2]
```diff
@@ -57,13 +57,46 @@ def mock_upload_trace_data():
...
+    """
+    store = mlflow.tracking._tracking_service.utils._get_store()
     with mock.patch("mlflow.tracking._tracking_service.utils._get_store") as mock_get_store:
-        mock_store = mock_get_store.return_value
-        mock_store.start_trace.side_effect = _mock_start_trace
-        mock_store.end_trace.side_effect = _mock_end_trace
-        yield mock_store
+        mock_get_store.return_value = store
+
+        _traces: Dict[str, TraceInfo] = {}
```
Comment (B-Step62) [inline tests/tracing/conftest.py:70]: note: A dirty way to make the mock stateful.

File: mlflow/tracking/client.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/tracking/client.py:878]: Isn't `mlflow.sourceRun` immutable too ?

Comment (B-Step62) [reply mlflow/tracking/client.py:878]: Nope, they are mutable indeed (search for `userImmutableRunTags` in universe).

File: mlflow/tracing/constant.py [hunk 3]
```diff
@@ -26,3 +27,6 @@ class SpanAttributeKey:
 
 # Trace request ID must have the prefix "tr-" appended to the OpenTelemetry trace ID
 TRACE_REQUEST_ID_PREFIX = "tr-"
+
+# A set of trace tags that cannot be updated by the user
+IMMUTABLE_TRACE_TAGS = {"mlflow.user", "mlflow.artifactLocation"}
```
Comment (harupy) [inline mlflow/tracing/constant.py:32]: Can we import `MLFLOW_USER` from `mlflow/utils/mlflow_tags.py`?

File: mlflow/tracing/processor/mlflow.py [hunk 4]
```diff
@@ -59,13 +60,20 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
 
     def _create_trace_info(self, span: OTelSpan) -> TraceInfo:
+        experiment_id = (
+            get_otel_attribute(span, SpanAttributeKey.EXPERIMENT_ID) or _get_experiment_id()
+        )
+        tags = {}
+
+        # If the span is started within an active MLflow run, we should record it as a trace tag
+        if mlflow.active_run():
+            tags[TraceTagKey.SOURCE_RUN] = mlflow.active_run().info.run_id
```
Comment (harupy) [inline mlflow/tracing/processor/mlflow.py:70]:
```suggestion
        if run := mlflow.active_run():
            tags[TraceTagKey.SOURCE_RUN] = run.info.run_id
```

**Review by harupy — APPROVED** [review]
Left a few comments, otherwise LGTM!

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

File: mlflow/tracing/constant.py (hunk 3 above)
Comment (B-Step62) [reply mlflow/tracing/constant.py:32]: makes sense!

**Pull request merged**